Optimistic locking

arthurclement-lab49 opened the pull request:
Hi,

I needed to have optimistic locking behavior on my project so implemented it in this pull request. It is based on a version field that gets incremented when saving. It can also be done through an updatedOn field but this PR uses the version technique for the DefaultModel.

It works the following way : 
A new field called "Version" has been added to DefaultModel (saves to "version" in the db)
Models have to implement a Versionable interface so that the behavior kicks in. DefaultModel implements it and so documents using the DefaultModel will have this enabled by default). This interface has 3 methods : 
GetVersion() -> returns the version number of this document
IncrementVersion() -> responsible for updating the version number
GetVersionFieldName() -> returns the name of the field that holds the version number

Then when the actual update is called, we can simply check that in the query part of the udpate that the version is the same as when the document was retrieved. If not, a custom error is raised.

arthurclement-lab49 committed:
add optimistic locking

arthurclement-lab49 committed:
add optimistic locking

arthurclement-lab49 committed:
add optimistic locking

arthurclement-lab49 committed:
handle adding add optimistic locking to documents created without it

arthurclement-lab49 committed:
handle adding add optimistic locking to documents created without it

mehran-prs commented:
Hi @arthurclement-lab49 
Thanks for your PR but the optimistic lock is a specific need and each user may want to implement it differently, it's not a generic feature for everyone, so maybe it's good to keep everything simple and let users implement it by themselves needs, for example, such a `DefaultModel`(that include the `version` field) is good for you, so maybe its good write it as a Custom `DefaultModel` in your project and use it as your `DefaultModel` in your models.

PRs that have more than 10line changes, it would be good to raise a feature request issue for them to talk about the feature before creating a PR, this will save your time.

I'll close this PR, feel free to reopen it if needed.

arthurclement-lab49 commented:
Hi @mehran-prs 

Thanks for looking into this. Can I just ask you look into my observations below before you definitely close this PR.

Optimistic locking is an essential feature for whatever project I have to implement and although mongodb now has transactions, you dont always want to use them, especially when updating a single document.
A project like https://mongoosejs.com/docs/guide.html#versionKey which is popular in the js world has it supported out of the box given it is such a common feature.

I appreciate that having it in baked in the defaultModel is perhaps too much as it will indeed add a version field that you might not want for you documents (although there is always the option of redefining your own DefaultModel as you mention).
A suitable alternative that you might consider then is to leave the VersionField struct in the code but not included in the default model. With this way, one would simply need to add the version field explicitely to his model to get optimistic locking in a simple manner : 
```
type User struct {
	mgm.DefaultModel `bson:",inline"`
	VersionField     `bson:",inline"`
}
```
The PR is actually flexible when it comes to this version field though as you just need to implement the Versionable interface.

But you also need to consider the most important part of the PR which is the update function of the operation.go file and will actually check whether the document you are saving has the right matching version.
Please consider at least this part of the PR as we can indeed work around the rest in the user land. Without this, users cannot just implement optimistic locking themselves unfortunately as the check needs to happen at the time of the update itself which is inside your library. If you dont want to merge it in, I and other users will have to do live with there own forks.
Thanks

mehran-prs commented on operation.go:
```diff
@@ -35,12 +38,44 @@ func update(ctx context.Context, c *Collection, model Model, opts ...*options.Up
 
-	res, err := c.UpdateOne(ctx, bson.M{field.ID: model.GetID()}, bson.M{"$set": model}, opts...)
+	query := bson.M{field.ID: model.GetID()}
+	modelVersionable, isVersionable := model.(Versionable)
```
What do you think about a little bit of change in variable names,...? (and maybe we should convert it to a method like `query,isVersionable:=updateVersion(m)`)

```golang
var v interface{}
vmodel, isVersionable := model.(Versionable)
if isVersionable{
v = modelVersionable.GetVersion()
if vmodel.IsVersionZero() {
    query["$or"] = bson.A{
    bson.M{vmodel.GetVersionFieldName(): v},
    bson.M{vmodel.GetVersionFieldName(): bson.M{"$exists": false}},
   }
} else {
	query[vmodel.GetVersionFieldName()] = v
}

vmodel.IncrementVersion()
}
```

arthurclement-lab49 replied:
No problem for variable names.
Not sure the additional method is really worth it, I personnally find it clearer without as its only meant to be used here at this stage.

mehran-prs commented on operation.go:
```diff
@@ -35,12 +38,44 @@ func update(ctx context.Context, c *Collection, model Model, opts ...*options.Up
 	}
 
+	if isVersionable && res.MatchedCount == 0 {
+		return fmt.Errorf("versioning error : document %v %v with version %v could not be found", c.Name(), model.GetID(), currentVersion)
```
Maybe it's good to convert it to a specific exported error  (e.g., `ErrVersionMismatch`), so users can check if the returned error is a version mismatch.

arthurclement-lab49 replied:
Ok. I've done it in the following way so we can keep a useful message for the user while still being able to use errors.Is(err, ErrVersionning) : 
```
var ErrVersionning = errors.New("versionning error")

...
return fmt.Errorf("document %v %v with version %v could not be found %w", c.Name(), model.GetID(), v, ErrVersionning)
...
```

mehran-prs commented:
Hi @arthurclement-lab49 
You are right. It's good to implement this feature in our library as it needs some changes in the library itself.
I have reviewed your PR and left some comments for you.
Thanks for your PR again.

arthurclement-lab49 commented on field.go:
```diff
@@ -18,6 +18,10 @@ type DateFields struct {
 }
 
+type VersionField struct {
+	Version int `json:"version" bson:"version"`
```
Yes that makes sense.
I think _v is a good choice indeed so will adjust that.

arthurclement-lab49 commented on model.go:
```diff
@@ -26,10 +26,17 @@ type Model interface {
 	SetID(id interface{})
 }
 
+type Versionable interface {
```
Yes good idea.

arthurclement-lab49 committed:
PR feedback : dont include version field by default, rename to _v, have specific error type

arthurclement-lab49 commented:
Hi mehran-prs

I've implemented your feedback. Its probably lacking a proper test at this stage to make sure everything works as expected. Will try to get to it later.

mehran-prs commented on field.go:
```diff
@@ -18,6 +18,10 @@ type DateFields struct {
 }
 
+type VersionField struct {
+	Version int `json:"_v" bson:"_v"`
```
What do you think about changing the struct field's name from `Version` to `V` too?

arthurclement-lab49 replied:
I dont think V is a very good name for a field. I would keep 'Version' personally given having VersionField is an opt in now and if you dont like it, you can always create your own VersionField. 
But if you really dont like Version, we could also use '_Version' which reflects nicely the serialized version '_v'.

mehran-prs replied:
The reason that I think it's good to change this field's name is the same as the reason that we changed tag values, I think users don't need to the Version field directly and just include it in their structs, so maybe it's good to let them have the `Version` name for their own fields.
`_Version` is good but it's private, I think a public field could be better.

arthurclement-lab49 replied:
Ok, I have changed it to Version_ which is public and also has the underscore to indicate it is a special field. Let me know what you think.
Also, dont forget we have removed VersionField from the default model so it is relatively easy to have your own VersionField defined with the name you want.

mehran-prs reviewed (changes requested):
Thanks for the changes.

mehran-prs commented:
Let's say a user implements `Versionable` using the Version field of type `time.Time`. 
When they create a document by calling it `Create()`, the version field value will be `0001-01-01 00:00:00` which is not what we wanted. I think it's good to set the document's first version at creation time by calling `doc.IncrementVersion` in the `Create` method. it makes sense for the version of type `int` too, as we know the first version of a document should be  `1` instead of zero.

After this change, we don't need to check if the version is zero in the `Update` method, but it's a good idea to keep that check to cover documents that already have been created without the optimistic lock feature.

arthurclement-lab49 committed:
PR feedback :
- Name version field name Version_ to avoid collisions.
- In operation.update(), capture Version before hooks in case updatedAt is used as versionable field
- In operation.create(), if model is versionable and zero, increment version so it gets initialized in DB
- Rename GetVersionFieldName to GetVersionBsonFieldName to make it clear it is the bson field name.
- Rename GetVersion() to Version() as it is more idiomatic

arthurclement-lab49 commented:
> Let's say a user implements `Versionable` using the Version field of type `time.Time`. When they create a document by calling it `Create()`, the version field value will be `0001-01-01 00:00:00` which is not what we wanted. I think it's good to set the document's first version at creation time by calling `doc.IncrementVersion` in the `Create` method. it makes sense for the version of type `int` too, as we know the first version of a document should be `1` instead of zero.
> 
> After this change, we don't need to check if the version is zero in the `Update` method, but it's a good idea to keep that check to cover documents that already have been created without the optimistic lock feature.
Yes ok, version will therefore start at 1 for the int strategy with this change which is ok, just need to be aware of it.

I've also done a couple of other changes as described in the last commit message : 

- Name version field name Version_ to avoid collisions. -> As discussed
- In operation.update(), capture Version before hooks in case updatedAt is used as versionable field -> Necessary if you want to use udpatedAt as the version field. I have thought more about this case and think that it would require even more changes if we wanted to support this case out of the box, eg have a struct DateFieldsVersionable that would implement the Versionable interface. Perhaps we can envisage this later but its worth keeping in mind.
- In operation.create(), if model is versionable and zero, increment version so it gets initialized in DB -> As discussed
- Rename GetVersionFieldName to GetVersionBsonFieldName to make it clear it is the bson field name. -> I think it is clearer
- Rename GetVersion() to Version() as it is more idiomatic -> go uses the simpler form for getters so I aligned with that.
